From f484f534165b4bc52add2e9c311ca011e4003602 Mon Sep 17 00:00:00 2001
From: Dennis Langenkamp <dennis.langenkamp@alliedvision.com>
Date: Wed, 23 Nov 2022 16:09:39 +0100
Subject: [PATCH] Fixed installation for yocto

---
 CMakeLists.txt | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 719df75..e086b99 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -25,5 +25,5 @@ include(V4L2ViewerLib.cmake)
 
 add_executable(V4L2Viewer Source/main.cpp)
 target_link_libraries(V4L2Viewer V4L2ViewerLib)
-install(TARGETS V4L2Viewer DESTINATION /)
-install(FILES LICENSE.md README.rst DESTINATION /)
+install(TARGETS V4L2Viewer)
+#install(FILES LICENSE.md README.rst DESTINATION /)
