From b4a73448692e52c4270a068eb3aac7afad495dcc Mon Sep 17 00:00:00 2001
From: Dennis Langenkamp <dennis.langenkamp@alliedvision.com>
Date: Wed, 23 Nov 2022 16:10:23 +0100
Subject: [PATCH] Use mmap by default, because the imx8 does not support
 userptr

---
 Source/Source/V4L2Viewer.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Source/Source/V4L2Viewer.cpp b/Source/Source/V4L2Viewer.cpp
index 685b107..a06f7c0 100644
--- a/Source/Source/V4L2Viewer.cpp
+++ b/Source/Source/V4L2Viewer.cpp
@@ -75,7 +75,7 @@ static int32_t int64_2_int32(const int64_t value)
 V4L2Viewer::V4L2Viewer(QWidget *parent, Qt::WindowFlags flags)
     : QMainWindow(parent, flags)
     , m_BLOCKING_MODE(true)
-    , m_BUFFER_TYPE(IO_METHOD_USERPTR) // use userptr by default
+    , m_BUFFER_TYPE(IO_METHOD_MMAP) // use userptr by default
     , m_NUMBER_OF_USED_FRAMES(5)
     , m_VIDIOC_TRY_FMT(true) // use VIDIOC_TRY_FMT by default
     , m_ShowFrames(true)
